src/Connection.php

```diff
@@ -1593,71 +1543,49 @@ public function createQueryBuilder(): QueryBuilder
      * @param array<int, int|string|Type|null>|array<string, int|string|Type|null> $types
      */
     final public function convertExceptionDuringQuery(
-        DriverException $e,
+        Driver\Exception $e,
         string $sql,
         array $params = [],
         array $types = []
-    ): Exception {
+    ): DriverException {
         $messageFormat = <<<'MESSAGE'
-An exception occurred while executing "%s"%s:
+An exception occurred while executing the query:
```

Comment:
Without the SQL being part of the exception message, this becomes just a static prefix. I don't think there are any other cases where a driver exception is instantiated.

What if we remove the $message argument from everywhere and instantiate the driver exception from the underlying exception and, optionally, a query.

Reply:
Makes sense. BTW,  if you really want to avoid leaking sensitive data, you must not repeat the underlying driver exception message that may contain some SQL:

```php
$connection->query("SELECT * 'sentitive data'");
```

> SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near **''sentitive data''** at line 1

Should we just always use a generic message like `'An exception occurred during query.'`? Not very informative, but secure.

Reply:
Let's take it one step at a time. I agree with the statement above but I don't want to creep the scope of this case.

Reply:
Basically, we have 2 ways to create driver exceptions:

- [Connection::convertExceptionDuringQuery()](https://github.com/doctrine/dbal/blob/4b8386c987e37f92c7e299279c0a5896e1bb4aa6/src/Connection.php#L1595)
- [Connection::convertException()](https://github.com/doctrine/dbal/blob/4b8386c987e37f92c7e299279c0a5896e1bb4aa6/src/Connection.php#L1622)

Should we just use 2 generic messages here:

- "An exception occurred during query"
- "An exception occurred in driver"

And *not* include the original exception message?

Reply:
> I agree with the statement above but I don't want to creep the scope of this case.

Sorry, not sure what you're suggesting?

Reply:
I'm suggesting to not make any changes outside the scope of this case. This case is about getting rid of JSON encoding by putting all the query information into the exception object instead of imprinting it in the message. A side effect of this is better handling of the potentially sensitive statement parameters. We can focus on improving it further in a separate case.

Reply:
Alright, so if I understand you correctly, you want to keep including the underlying exception message in the `DriverException` message?

Reply:
Correct. But I'm not sure if the static prefix is needed any longer. Could we just pass the underlying exception to the `DriverException` and make it extract the message and assign it to itself?

Reply:
Sure we can. It all depends on what you want in this exception message. If you really want to make it bulletproof secure, you just put the prefix and nothing else. Otherwise the original exception message, with or without the prefix.

A couple thoughts, unrelated to each other:

- Even if you want to keep the prefix, you don't need to pass it down the `ExceptionConverter`, you can just hardcode it in the `DriverException` constructor, and choose which prefix to use depending on whether `$query` is set;
- TBH I'm not sure how insecure putting sensitive data in the exception message is; we should advocate best practices and not displaying exception messages anyway. And if you're using a third-party error reporting system, you'll usually trust it not to share sensitive data it might intercept. My two cents.

Reply:
> And if you're using a third-party error reporting system, you'll usually trust it not to share sensitive data it might intercept. My two cents.

It's not about trust. E.g. according to GDPR you should be able to erase customer data from the database and you have a procedure for that. If this data lands in the logs, you should be able to process your logs as well which is absolutely not necessary.

Reply:
Since it's not obvious how to proceed in this case, let's leave it as is.

Reply:
> You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ''sentitive data'' at line 1

I'm curious if the same is reproducible with a syntactically valid query. A syntax error is an indicator of something utterly broken in the code. Allowing such code to deal with sensitive data and blaming logging for leaking it is not the best strategy.

Reply:
> Since it's not obvious how to proceed in this case, let's leave it as is.

I pushed 1b0820032ab75e6a2d70680f202cb93dfc620f1a to test creating the exception message in `DriverException`, IMO it looks much cleaner. I can revert it if you wish, though.

Reply:
No, it looks good.